providers/enchant_nuspell.cpp

```diff
@@ -0,0 +1,438 @@
+		return false;
+	if (strncmp (dir_entry, tag, tag_len) != 0)
+		return false;
+	//e.g. requested dict for "fi",
+	//reject "fil_PH.dic"
+	//allow "fi-FOO.dic", "fi_FOO.dic", "fi.dic", etc.
+	if (!ispunct(dir_entry[tag_len]))
+		return false;
+	return true;
+}
+
+static char *
+nuspell_request_dictionary (const char * tag)
+{
+	vector<string> names;
+
+	s_buildHashNames (names, tag);
+
+	for (size_t i = 0; i < names.size (); i++) {
+		if (g_file_test(names[i].c_str(), G_FILE_TEST_EXISTS) &&
+		    s_fileExists(s_correspondingAffFile(names[i]))) {
+			return strdup (names[i].c_str());
+		}
+	}
+
+	vector<string> dirs;
+	s_buildDictionaryDirs (dirs);
+
+	for (size_t i = 0; i < dirs.size(); i++) {
+		GDir *dir = g_dir_open (dirs[i].c_str(), 0, nullptr);
+		if (dir) {
+			const char *dir_entry;
+			while ((dir_entry = g_dir_read_name (dir)) != NULL) {
+				if (is_plausible_dict_for_tag(dir_entry, tag)) {
+					char *dict = g_build_filename (dirs[i].c_str(),
+								       dir_entry, nullptr);
+					if(s_fileExists(s_correspondingAffFile(dict))) {
+						g_dir_close (dir);
+						return dict;
+					}
+					g_free(dict);
+				}
+			}
+
+			g_dir_close (dir);
+		}
+	}
+
+	return NULL;
+}
+
+bool
+NuspellChecker::requestDictionary(const char *szLang)
+{
+	char *dic = nuspell_request_dictionary (szLang);
+	if (!dic)
+		return false;
+	string aff(s_correspondingAffFile(dic));
+	if (!s_fileExists(aff))
+		return false;
+	auto path = string(dic);
+	if (path.size() >= 4 && path.compare(path.size() - 4, 4, ".dic") == 0)
+		path.erase(path.size() - 4);
+	else
+		return false;
+	try {
+		nuspell = nuspell::Dictionary::load_from_path(path);
+	} catch (const std::runtime_error& e) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Enchant
+ */
+
+static char **
+nuspell_dict_suggest (EnchantDict * me, const char *const word,
+		      size_t len, size_t * out_n_suggs)
+{
+	NuspellChecker * checker = static_cast<NuspellChecker *>(me->user_data);
+	return checker->suggestWord (word, len, out_n_suggs);
+}
+
+static int
+nuspell_dict_check (EnchantDict * me, const char *const word, size_t len)
+{
+	NuspellChecker * checker = static_cast<NuspellChecker *>(me->user_data);
+
+	return !(checker->checkWord(word, len));
+}
+
+static int
+nuspell_dict_is_word_character (EnchantDict * me _GL_UNUSED_PARAMETER,
+				uint32_t uc, size_t n _GL_UNUSED_PARAMETER)
+{
+	return g_unichar_isalpha(uc);
+}
+
+static void
+nuspell_provider_enum_dicts (const char * const directory,
+			     vector<string> & out_dicts)
+{
+	GDir * dir = g_dir_open (directory, 0, nullptr);
+	if (dir) {
+		const char * entry;
+		while ((entry = g_dir_read_name (dir)) != NULL) {
+			char * utf8_entry = g_filename_to_utf8 (entry, -1, nullptr, nullptr, nullptr);
+			if (utf8_entry) {
+				string dir_entry (utf8_entry);
+				g_free (utf8_entry);
+
+				int hit = dir_entry.rfind (".dic");
+				if (hit != -1) {
+					/* don't include hyphenation dictionaries
```

Comment:
This comment would be clearer split into two: first, above `if (hit != -1)`, "Don't include hyphenation dictionaries", and then in the current position, "Require .aff file to be present".

Reply:
See https://github.com/nuspell/enchant/commit/35cdcaaf511f8c84ff8a66b9aafca22699b78253#diff-b4a4b7f548e6522299f21f4f5632ff1cR311